compiler/modules/front/const.bal

```diff
@@ -544,47 +547,63 @@ function foldVarRefExpr(FoldContext cx, t:SemType? expectedType, s:VarRefExpr ex
     }
 }
 
-function foldFloatLiteralExpr(FoldContext cx, t:SemType? expectedType, s:FpLiteralExpr expr) returns s:ConstValueExpr|FoldError {
-    // This will need to change when we support decimal
-    float|error result = floatFromDecimalLiteral(expr.untypedLiteral);
-    if result is float {
-        return { startPos: expr.startPos, endPos: expr.endPos, value: result };
+function foldFpLiteralExpr(FoldContext cx, t:SemType? expectedType, s:FpLiteralExpr expr) returns s:ConstValueExpr|FoldError { 
+    var { typeSuffix, untypedLiteral, startPos } = expr;
+    float|decimal result;
+    if typeSuffix != () {
+        result = typeSuffix is s:FLOAT_TYPE_SUFFIX ? floatFromFpLiteral(untypedLiteral) : check decimalFromFpLiteral(cx, untypedLiteral, startPos);
+    }
+    else if expectedType == () || t:includesSome(expectedType, t:FLOAT) || !t:includesSome(expectedType, t:DECIMAL) {
+        result = floatFromFpLiteral(untypedLiteral);
     }
     else {
-        return cx.semanticErr("invalid float literal", cause=result, pos=expr.startPos);
+        result = check decimalFromFpLiteral(cx, untypedLiteral, startPos);
     }
+    return { startPos: startPos, endPos: expr.endPos, value: result };
 }
 
 function foldIntLiteralExpr(FoldContext cx, t:SemType? expectedType, s:IntLiteralExpr expr) returns s:ConstValueExpr|FoldError {
-    float|int|error result;
-    string ty;
-    if expr.base == 10 && expectsFloat(expectedType) {
-        result = floatFromDecimalLiteral(expr.digits);
-        ty = "float"; 
+    Position startPos = expr.startPos;
+    int|float|decimal result;
+    if expectedType == () || t:includesSome(expectedType, t:INT) {
+        result = check intFromLiteral(cx, expr);
+    }  
+    else if t:includesSome(expectedType, t:FLOAT) {
+        result = floatFromFpLiteral(expr.digits);
     }
-    else {
-        result = s:intFromIntLiteral(expr.base, expr.digits);
-        ty = "int";
-    }
-    if result is int|float {
-        return { startPos: expr.startPos, endPos: expr.endPos, value: result };
+    else if t:includesSome(expectedType, t:DECIMAL) {
+        result = check decimalFromFpLiteral(cx, expr.digits, startPos);
     }
     else {
-        return cx.semanticErr("invalid " + ty + " literal", cause=result, pos=expr.startPos);
+        result = check intFromLiteral(cx, expr);
     }
+    return { startPos: startPos, endPos: expr.endPos, value: result };
 }
 
-function expectsFloat(t:SemType? semType) returns boolean {
-    if semType == () {
-        return false;
-    }
-    else {
-        return t:isSubtypeSimple(t:intersect(semType, t:union(t:FLOAT, t:INT)), t:FLOAT);
+// Since the binary floating point literal is parsed correctly,
+// it is impossible to throw an error.
```

Comment:
We don't "throw" errors in Ballerina. We either panic or return errors. (Applies to next two functions also.)

Reply:
Changed